### ISSUE TITLE
False N+1 query report for object instance (not query)

### ISSUE THREAD

**Issue by IgorMarques** [flyerhzm/bullet#239]
Hello, guys,

I'm using Ruby 2.2. with Rails 4.2 and Bullet 4.14.4 to build an app. At one of my views, I got the following message:

> N+1 Query detected Politician => [:account] Add to your finder: :includes => [:account]
> 
> N+1 Query method call stack
> 
> app/models/user.rb:19:in `account'
> 
> app/controllers/home_controller.rb:6:in `index'

Here is my action at home controller:

``` ruby
@account = current_user.account
@new_contacts = current_user.contacts.created_this_week.count
@new_collabs = current_user.collaborators.created_this_week.count
```

And the relevant section of user model:

``` ruby
belongs_to :role, polymorphic: true, dependent: :destroy
delegate :account, to: :role
delegate :collaborators, to: :role
delegate :contacts, to: :role
```

I already tried to the following:

``` ruby
@account = current_user.account.includes(:contacts, :collaborators)
```

But then I only get the error:

> undefined method `includes' for Account:0x007fa7474e04a8

I did some research and what it seems is that includes works only for relations (which is not the case).

Is bullet giving a false positive?

**Comment by kamsz** [comment]
I am experiencing the same problem.

**Comment by mordocai** [comment]
Also seeing this. Can we confirm that this is indeed a bug in bullet? This is causing CI failures so if it is it'd be nice to have a fix.

**Comment by flyerhzm** [comment]
anyone can build a sample app for me to reproduce the issue?

**Comment by mordocai** [comment]
Working on one now

**Comment by mordocai** [comment]
My naive attempt at reproducing did not work. Will work on it again on monday unless someone else figures it out before then.

**Comment by malroc** [comment]
Same here. Bullet reports N + 1 query for a SINGLE object (not collection).
Environment: `devise` + `cancancan` + `bullet`

Codebase (simplified):


    class User
      has_many :conferences, foreign_key: :owner_id
    end

    class Conference
      belongs_to :owner, class_name: "User"
    end

    class ConferencesController < ApplicationController
      load_resource :user, find_by: :login
      load_and_authorize_resource find_by: :slug, through: :user
    
      def create
        @conference.save
        respond_with @user, @conference
      end
    end

Bullet output:

    N+1 Query detected
      Conference => [:owner]
      Add to your finder: :includes => [:owner]
    N+1 Query method call stack
      /Users/okalistratov/Documents/sites/devconf/app/controllers/conferences_controller.rb:23:in `create'

Line 23 is `respond_with @user, @conference`
`owner` association is called when generating response, but that really doesn't matter since `conference` is a single object, not a collection, and cannot generate N + 1 query.
Collection association is called by `cancancan` when executing `load_ant_authorize_resource`, the resulting loading sequence would look like:

    @user = User.find_by_login(params[:user_id])
    @conference = @user.conferences.find_by_slug(params[:slug)

+ some authorization code that shouldn't matter.
I.e. there is an intermediate collection association here, but it is not used directly, and the resulting object is not a collection.

**Comment by jguitar** [comment]
Are there some news about this?

**Comment by flyerhzm** [comment]
seems I can't reproduce it.

**Comment by Uysim** [comment]
I meet the same problem

**Comment by printercu** [comment]
@flyerhzm I've made an example app here https://github.com/printercu/bullet_example

**Comment by printercu** [comment]
Seems like #308 is related to this one

**Comment by flyerhzm** [comment]
@printercu I have fixed your scenario in bullet 5.4.3, please check.

**Comment by printercu** [comment]
Warning disappeared. Thanks a lot!
